rust/kernel/fs.rs

```diff
@@ -464,17 +495,19 @@ pub struct SuperParams {
 
 /// A superblock that is still being initialised.
 ///
+//// It uses type states to ensure that callers use the right sequence of calls.
+///
 /// # Invariants
 ///
 /// The superblock is a newly-created one and this is the only active pointer to it.
-pub struct NewSuperBlock<'a, T: FileSystem + ?Sized> {
+pub struct NewSuperBlock<'a, T: FileSystem + ?Sized, S = NeedsInit> {
```

Comment:
I forgot, did you already try to merge `SuperBlock` and `NewSuperBlock`? Did that not work?

Reply:
It didn't work for a couple of reasons:
1. `SuperBlock` is a transparent type to a C struct. Rust never really has a `SuperBlock` itself, it's always a reference to one. So I was having a hard time "transferring ownership" or "consuming" when, which is something I need to do to transition from one typestate to another (otherwise one could transition more than once).
2. I wanted to attach an invariant lifetime to a new superblock to guarantee that the one returned by `fill_super` was in fact the one passed in. But I didn't want to add a "phantom" one to `SuperBlock` because that would pollute it everywhere and I don't want it.

I ended up implementing `Deref<Target = SuperBlock>` for the `NeedsRoot` typestate.

If you have ideas on how to work around the above, I'd be interested in trying them out.

Reply:
I am wondering if it really is necessary to have this complex initialization process. Do you have some examples (maybe just some C code) of what needs to be possible (do you think that it's a good idea to look at puzzle fs for this)? From what `tarfs` needs I think it could be done with a couple of callbacks for creating data and creating the root.

Reply:
puzzlefs doesn't deal with block devices, so it doesn't have to deal with initialising the block size before it can read it.

If you run `git grep fill_super`, you'll find a bunch of them. (This is a "new api" so not all C file systems have been converted yet.)

The reason for this complex init at the moment is to allow a single linear `fill_super` where all necessary state to initialise the superblock is present in a single function. It may not be worth it though and this is the most complicated part of the series.

One thing that Kent suggested, for example, was to use something other than `sb_bread` to read the super-block itself (or whatever blocks would be needed during fill_super). That would obviate the need to separate the block_size init and reading blocks during init.

We still want to separate superblock data init and creating inodes because inodes give access to the superblock data (via `inode.super_block().data()`). So perhaps two callbacks: `init_super` and `get_root` would suffice to reduce the complexity here.

What did you have in mind?

Reply:
Remove the type states. Now we have two callbacks: `super_params` and `init_root`.